### ISSUE TITLE
while_let false positive

### ISSUE THREAD

**Issue by lrbalt** [Manishearth/rust-clippy#1017]
I have this loop:
```rust
loop {
   match reader.read(&mut buf) {
      Err(_) => thread::sleep(Duration::from_millis(500)),
      Ok(length) => {
         len = length;
         break
      }
   }
}
```
I need to use len in the code that follows. Clippy suggests to use `while let Err(_) = ...` construct, but that doesn't let me use 'length' in 'Ok(length)'

**Comment by lrbalt** [comment]
perhaps clippy can check on `Ok(_)` versus `Ok(variable)`

**Comment by mcarton** [comment]
This is a bug, we need to check that the type is `Option`.

**Comment by mcarton** [comment]
Thanks for the report! Turns out Clippy was disturbed by the colon-less `break`.

**Comment by lrbalt** [comment]
Thanks!
